tests/explicit-nulls/neg/flow6.scala

```diff
@@ -1,70 +1,50 @@
-
-// Test that flow inference behaves soundly within blocks.
-// This means that flow facts are propagated to ValDefs, but not
-// to DefDefs or lazy vals or implicit vals.
-// The reason is that forward references are allowed for `defs`,
-// so when we see
-// ```
-// f
-// if (x == null) return "foo"
-// def f = x.length
-// ```
-// we can't assume that f is executed _after_ the if statement
-// is executed.
+// Test forward references handled with flow typing
+// Currently, the flow typing will not be applied to definitions forwardly referred.
 class Foo {
 
   def test1(): Unit = {
     val x: String|Null = ???
     if (x == null) return ()
-    val y = x.length // ok: x: String inferred
+    def y = x.length // ok: x: String inferred
     ()
   }
 
+  // This test is similar to test1, but a forward DefDef referring
+  // to y is added after x. y is completed before knowing the
+  // fact "x != null", hence, the type x: String|Null is used.
   def test2(): Unit = {
     val x: String|Null = ???
+    def z = y
     if (x == null) return ()
-    def y = x.length // error: x: String|Null inferred
+    def y = x.length // error: x: String|Null is inferred
     ()
   }
 
+  // Since y is referred before definition, flow typing is not used here.
   def test3(): Unit = {
     val x: String|Null = ???
+    lazy val z = y
     if (x == null) return ()
-    lazy val y = x.length // error: x: String|Null inferred
+    lazy val y = x.length // error: x: String|Null is inferred
     ()
   }
 
+  // This case is not valid but the test1 above is, because
+  // non-lazy value definitions exist between forward references.
+  // Since y is referred (by z) before definition, flow typing is not used here.
+  // Only the typing error is shown because reference check is after typing.
   def test4(): Unit = {
+    val z = implicitly[Int]
     val x: String|Null = ???
     if (x == null) return ()
     implicit val y: Int = x.length // error: x: String|Null inferred
   }
 
-  // This case is different from #3 because the type of y doesn't need
-  // to be inferred, which triggers a different codepath within the completer.
+  // Since z is referred before definition, flow typing is not used here.
   def test5(): Unit = {
     val x: String|Null = ???
     if (x == null) return ()
-    lazy val y: Int = x.length // error: x: String|Null inferred
-    ()
-  }
-
-  def test6(): Unit = {
-    val x: String|Null = ???
-    if (x == null) return ()
-    def y: Int = x.length // error: x: String|Null inferred
-    ()
-  }
-
-  // This test checks that flow facts are forgotten for defs, but only
```

Comment:
@noti0na1 could you please restore this test?

Reply:
The old tests in `neg/flow6.scala` have been moved to `pos/flow6.scala`